### DISCUSSION TITLE
FLTK install question, completely lost

### DISCUSSION THREAD

**Discussion by javakotlinuser** [fltk/fltk#1013] [Q&A]
## help sincerely appreciated
I have absolutely 0 idea how to install FLTK. I've been trying for my entire day now, I downloaded fltk-1.3.9-source.tar.gz, uncompressed it and put it in a folder called "C:\C++ LIBRARIES". I tried to run a simple program but the #include could never work correctly, I have tried looking on Youtube, But what I found was: Extremely outdated tutorials, and tutorials for operating systems I haven't even heard of, I don't have Visual Studio 2022, nor the space for it, but instead have Visual Studio Code, I run Windows 11.
## Code I tried running:
```C++
#include <FL/Fl.H>
#include <FL/Fl_Box.H>
#include <FL/Fl_Window.H>

int main() {
  Fl_Window window(200, 200, "Window title");
  Fl_Box box(0, 0, 200, 200, "Hello, World!");
  window.show();
  return Fl::run();
}
```

**Comment by Albrecht-S** [comment]
Hi @javakotlinuser, as the name of the download file says, it's **_source code_** only, we don't provide pre-built libraries for FLTK. This means that you need to compile and build the FLTK libraries after you downloaded and expanded the tarball. To do this you need a build tool like Visual Studio, MinGW, or MSYS2 on Windows. Visual Studio Code alone is not sufficient: it's a good code editor with some extensions but not a build tool.

If this is all new to you, then your best choice is probably to download and install the free version of Visual Studio (VS) Community (aka Visual Studio Express) but if you don't have the space to download VS you can't build FLTK on your system. Maybe you can use another system with more free disk space?

The other choices (MinGW or MSYS2) may be smaller to install but are somewhat more complicated to use.

In the downloaded FLTK sources you can find two README files of interest (among others):
- `README.MSWindows.txt` on how to build FLTK and your own program (¹)
- `README.CMake.txt` on how to use `CMake` to create the build files you need for Visual Studio (²).

I'm sorry if this all sounds cryptic to you, but this is how building a library (FLTK) and your program from source code works. (FYI: This is different than other programming languages like `Basic` or `Python` that don't need this build step because they are interpreted rather than compiled.)

------
(¹) FYI: in the upcoming release 1.4.0 this file is called `README.Windows.txt`
(²) FLTK 1.3.x comes with bundled IDE files for Visual Studio, thus you may not need this one and don't need `CMake` at all but this will be required in the next version (1.4.0).

**Reply by javakotlinuser** [reply]
Ok I broke down and got Visual Studio, I was discouraged in the first place because it looks complicated to use, I already had MinGW and MYSYS though already as that's how I got GCC to compile my code, Might I ask what Visual Studio addons i might need?

**Reply by Albrecht-S** [reply]
1. If you have MinGW and/or MSYS2 installed you can use it to build FLTK and your app.
2. The only optional part of Visual Studio I'm aware of that you need to install is called something like "C++ for Desktop Apps". Note that this is off the top of my head, you'll hopefully see what it's really called when installing VS. This is needed because FLTK is used to build Windows Desktop apps with a C++ compiler.

**Comment by MoAlyousef** [comment]
Also note that visual studio code, unlike visual studio, doesn’t bundle a C++ compiler. 
You have a few options of getting a compiler on Windows; if you don’t want to install visual studio and the msvc toolchain, I would recommend installing msys2 which offers gcc via its package manager. It also provides prebuilt fltk via its package manager.

**Reply by javakotlinuser** [reply]
I've been having GCC and G++ via MinGW via MYSYS2

**Reply by MoAlyousef** [reply]
You can try the following:
1- Open an msys2 shell.
2- enter the following commands:
```
source shell mingw64
pacman -S $MINGW_PACKAGE_PREFIX-fltk $MINGW_PACKAGE_PREFIX-cmake
```
3- Create a project with a CMakeLists.txt:
```cmake
cmake_minimum_required(VERSION 3.15)
project(app)

find_package(FLTK CONFIG REQUIRED)

add_executable(main main.cpp)
target_link_libraries(main PRIVATE fltk)
```
4- Create a minimal main.cpp with some fltk code.
5- Configure and build (cmake -Bbin && cmake --build bin)
